### PR TITLE
Translatable tile names and descriptions.

### LINKED ISSUE #8: Tile localization and better tooltips
In an effort to improve the tilesheet extension the following plan has been laid out which will allow for translation of tilesheet icons as well as improved tooltips. Even translation into English will be useful as it will allow multiple icons on a tilesheet to share the same display name, as well as allowing custom tooltips to be assigned across the wiki.

* [x] Add a table to the database with 4 columns for entry id (matching the entry id in the existing table), language code (such as `en`, `de`, or `ru`), display name, and description.
* [x] Add a special page where translators can translate tilesheet icons, assigning them a display name and description that suits their language.
* [x] For invoking the tilesheet extension add tooltip parameters and having the extension output the appropriate tooltip code, instead of the wrapping template handle it. Order of priority for the tooltip is tooltip parameter at invocation, display name + description from the new table, and defaulting to the item name in the tilesheet itself.

### PR DESCRIPTION
**Pull request by elifoster**
Tiles can now have their display names and descriptions translated

* Add a new parser function #iconloc which takes the item and mod names, the type, and the language, and returns the localized data for that type (name or description) and language.
* Add a new special page TileTranslator for translating existing translations, and creating new ones. A link to this page for each entry is added to the TileList table, just like the Edit link.
* Fix a couple issues with CreateTileSheet that are somewhat unrelated, but prevented me from testing.
* Add an entirely new table, ext_tilesheet_languages, which contains the columns entry_id (not a primary key), lang, display_name, and description. entry_id is identical to that in ext_tilesheet_items. lang contains language codes, and display_name and description are self explanatory. entry_id is not a primary key because we need it to be able to be duplicated for multiple different langauges.
The SQL install code will populate the language table with all of the existing item entries, defaulting the language to en and the name to the item_name.

Close #8

@retep998 and @xbony2 I'd like your feedback on this.

Note: After this is merged, I will still need to update the changelog and version numbers before it gets deployed!

### CODE REVIEW

File: i18n/en.json [hunk 1]
```diff
@@ -63,6 +64,18 @@
...
+	"tilesheet-tile-translator-description": "Description",
+	"tilesheet-tile-translator-language": "Language Code",
+	"tilesheet-tile-translator-language-hint": "Change the language code to create a new translation with the above values.",
+	"tilesheet-tile-translator-submit": "Update",
+	"tilesheet-tile-translator-filter-id": "Entry ID",
+	"tilesheet-tile-translator-filter-language": "Language Code",
+	"tilesheet-tile-translator-filter-legend": "Search",
+	"tilesheet-tile-translator-filter-submit": "Go",
+	"tilesheet-tile-translator-legend": "Translate",
+	"tilesheet-tile-translator-delete": "<span style=\"color:red; font-weight:bold;\">Delete Translation</span>",
```
Comment (xbony2) [inline i18n/en.json]: Does HTML shit really need to be in the localizations file?

**Comment by xbony2** [comment]
Web API is wanted for this, since potentially that's how all of the translations will be imported.

File: i18n/en.json (hunk 1 above)
Comment (retep998) [inline i18n/en.json]: It doesn't. Blame santa.

**Comment by retep998** [comment]
@xbony2 There is already a translate link next to each tile in the tile list which you can click on to translate that tile to different languages.

**Comment by xbony2** [comment]
Manually? I don't want translators/me spending all of their time translating the tile names. There are hundreds of thousands of tiles; it's a time-waster.

**Comment by elifoster** [comment]
I agree. I'll implement an API.

File: i18n/en.json (hunk 1 above)
Comment (elifoster) [inline i18n/en.json]: No, but it's in the other ones and I wanted to remain consistency. When I end up fixing it, I will fix all of the cases.

**Comment by elifoster** [comment]
I implemented editing APIs for this. @xbony2 I think I should also make a query API:
* Provide an ID and a language, return the name and description. Example:
```
{
  'name': 'Pangalan',
  'description': 'Ito bagay sucks'
}
```
* Provide just an ID, return an array of language, name, and description. Example:
```
[
  {
    'name': 'Pangalan',
    'description': 'Ito bagay sucks',
    'language': 'tl'
  },
  {
    'name': 'Nombre',
    'description': 'Eso objecto es asco'
    'language': 'es'
  }
]
```

what do you guys think of this?

**Comment by xbony2** [comment]
Could be useful.

**Comment by elifoster** [comment]
There. How does this PR, as a whole, look now @retep998 and @xbony2 ?

File: api/TilesheetsDeleteTranslationApi.php [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        );
+    }
+
+    public function getDescription() {
+        return 'Deletes a translation for an entry ID.';
+    }
+
+    public function getExamples() {
+        return array(
+            'api.php?action=deletetranslation&tsid=6&tslang=es-ni',
```
Comment (xbony2) [inline api/TilesheetsDeleteTranslationApi.php]: Bad example if someone plans to use this on other wikis.

File: special/TileTranslator.php [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+        $form .= TilesheetsForm::createFormRow('tile-translator-filter', 'id', $id, 'number', 'min="1" id="form-entry-id"');
+        $form .= TilesheetsForm::createFormRow('tile-translator-filter', 'language', $language);
+        $form .= TilesheetsForm::createSubmitButton('tile-translator-filter');
+        $form .= "</table>";
+
+        $out = Xml::openElement('form', array('method' => 'get', 'action' => $wgScript, 'id' => 'ext-tilesheet-tile-translator-filter')) .
+            Xml::fieldset($this->msg('tilesheet-tile-translator-filter-legend')->text()) .
+            Html::hidden('title', $this->getPageTitle()->getPrefixedText()) .
+            $form .
+            Xml::closeElement( 'fieldset' ) . Xml::closeElement( 'form' ) . "\n";
```
Comment (xbony2) [inline special/TileTranslator.php]: Putting a space between ( and ' is the worst convention I've ever seen, and you better be ashamed you did it by accident.

Comment (elifoster) [inline special/TileTranslator.php]: This method was mostly copied from another special page, so it doesn't surprise me that that's how it is.

[Also, side note, MediaWiki actually recommends this trash style](https://www.mediawiki.org/wiki/Manual:Coding_conventions/PHP)

Comment (xbony2) [inline special/TileTranslator.php]: No one who ever worked for MW became a millionaire, and this is why. Except for the ones who ditched, of course.

**Comment by elifoster** [comment]
@Alexia could you perform a code review on this?

File: install/sql/ext_tilesheet_languages.sql [hunk 4]
```diff
@@ -0,0 +1,6 @@
+CREATE TABLE IF NOT EXISTS ext_tilesheet_languages (
+  `entry_id` int(11) NOT NULL,
+  `lang` varchar(10) NOT NULL,
+  `display_name` varchar(100),
+  `description` varchar(100)
+);
```
Comment (Alexia) [inline install/sql/ext_tilesheet_languages.sql]: This line needs the table options.  Add this comment to have Mediawiki substitute it.
`) /*$wgDBTableOptions*/;`

File: install/sql/ext_tilesheet_languages.sql [hunk 5]
```diff
@@ -0,0 +1,6 @@
+CREATE TABLE IF NOT EXISTS ext_tilesheet_languages (
```
Comment (Alexia) [inline install/sql/ext_tilesheet_languages.sql]: Prepend the table name with `/*_*/` so Mediawiki can substitute table prefixes.

**Comment by Alexia** [comment]
I did not see any security issues.  Just the issues with the SQL install.

File: install/sql/ext_tilesheet_languages.sql (hunk 4 above)
Comment (retep998) [inline install/sql/ext_tilesheet_languages.sql]: The `entry_id` should definitely be an index since we're going to do a lot of lookups by it.

**Comment by elifoster** [comment]
I have resolved those issues with the sql install

**Pull request merged**